integrate ngfx backend

jmoguillansky-gpsw opened the pull request.

jmoguillansky-gpsw commented:
Current status: 282/296 (95%) passing (ngfx/d3d12)

Summary of Failures:

29/296 ngl-tests:ngfx+benchmark / fingerprint_without_compute FAIL 25.72s exit status 1
30/296 ngl-tests:ngfx+benchmark / resources_without_compute FAIL 10.85s exit status 1
31/296 ngl-tests:ngfx+benchmark / fingerprint_with_compute FAIL 1.60s exit status 1
32/296 ngl-tests:ngfx+benchmark / resources_with_compute FAIL 1.51s exit status 1
56/296 ngl-tests:ngfx+compute / animation_post_render FAIL 1.37s exit status 1
57/296 ngl-tests:ngfx+compute / histogram FAIL 3.11s (exit status 2170 or signal 2042 SIGinvalid)
58/296 ngl-tests:ngfx+compute / image_load_store FAIL 2.80s (exit status 2170 or signal 2042 SIGinvalid)
59/296 ngl-tests:ngfx+compute / particles FAIL 0.96s exit status 1
207/296 ngl-tests:ngfx+media / timeranges_rtt FAIL 1.38s exit status 1
218/296 ngl-tests:ngfx+rtt / load_attachment FAIL 1.15s exit status 1
219/296 ngl-tests:ngfx+rtt / load_attachment_nested FAIL 1.16s exit status 1
227/296 ngl-tests:ngfx+rtt / load_attachment_msaa FAIL 1.14s exit status 1
228/296 ngl-tests:ngfx+rtt / load_attachment_nested_msaa FAIL 1.14s exit status 1
252/296 ngl-tests:ngfx+shape / triangles_mat4_attribute FAIL 2.73s (exit status 2170 or signal 2042 SIGinvalid)

Ok: 282
Expected Fail: 0
Fail: 14
Unexpected Pass: 0
Skipped: 0
Timeout: 0

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -64,17 +64,30 @@
     ),
+    # TODO: push change upstream to shaderc, or use G-P-S repo
+    shaderc=dict(
+        url="https://github.com/jmoguillansky-gpsw/shaderc.git",
```
TODO: use shaderc from upstream

jmoguillansky-gpsw replied:
TODO: do we need to add shaderc here, or will ngfx pull shaderc?

wmacdonald-gpsw commented on libnodegl/backends/ngfx/gpu_ctx_ngfx.h:
```diff
@@ -0,0 +1,66 @@
+        rendertarget *rt = nullptr, *rt_load = nullptr;
+    } offscreen_resources;
+
+    texture *dummy_texture = nullptr;
```
Can you clarify the purpose of the dummy_texture?

jmoguillansky-gpsw replied:
I'm not 100% sure, it's also used in the vulkan backend.  It might be used as part of ngl-probe to probe which GPU contexts are available. Need to ask Matthieu...

jmoguillansky-gpsw commented on libnodegl/meson.build:
```diff
@@ -525,7 +525,7 @@ foreach gbackend_name, gbackend_cfg : gbackends_cfg
         endforeach
         glslang_ver = '.'.join(glslang_ver_array)
-        glslang_min_ver = '>= 11.3.0'
+        glslang_min_ver = '>= 11.3'
```
on my pc, when I tried to build, the glslang_min_ver was returning 11.1 instead of 11.1.0, and the condition '>= 11.1.0' was returning false in meson (it's using a basic string comparison, not a number comparison), and then meson didn't include this backend.  When I changed it to '>= 11.1' it worked. 
When I rebased I updated to version 11.3

jmoguillansky-gpsw commented on pynodegl-utils/pynodegl_utils/tests/cmp.py:
```diff
@@ -55,10 +55,10 @@ def compare_data(self, test_name, ref_data, out_data):
         test_tmpdir = op.join(get_nodegl_tempdir(), "tests")
         os.makedirs(test_tmpdir, exist_ok=True)
-        filename = op.join(test_tmpdir, f"{func_name}_{dump_index:03}.png")
+        filename = op.join(test_tmpdir, f"{func_name}_{dump_index:03}_{backend}.png")
```
These dumped files are for debugging only, they're not referenced for validating unit tests, etc.  Before this change, different backends would overwrite these files, and we didn't know which backend generated these files, etc.  Now we can compare side by side, and know which backend generated the file, etc.

ylaala-gpsw commented on libnodegl/pgcraft.c:
```diff
@@ -1354,27 +1354,43 @@ static void setup_glsl_info_gl(struct pgcraft *s)
 
-    s->has_explicit_bindings        = 1;
-    s->has_in_out_qualifiers        = 1;
+    s->has_explicit_bindings = 1;
```
Make sure to limit diff by not changing format otherwise make it in a separate commit but I think this is just a style preference that should not be done.

jmoguillansky-gpsw replied:
updated

ylaala-gpsw commented on libnodegl/meson.build:
```diff
@@ -729,9 +793,17 @@ run_target(
+  # add_global_arguments doesn't seem to work here
+  gen_doc_cpp_args = []
+  if cc.get_id() == 'msvc'
+    gen_doc_cpp_args += ['/std:c++20']
```
Is this a requirement? It feel that should be define by the user environment / driven by the client

jmoguillansky-gpsw replied:
I'm using a very small feature of C++20 in ngfx.  If necessary we can switch back to C++17.  
Specifically I'm using the "designated initializers" for initializing structs (something C always supported), e.g.:
my_struct s = {
  .param0 = ...,
  .param1 = ...
}

jmoguillansky-gpsw replied:
I called add_global_arguments meson function to set global c++ args but it didn't seem to work correctly, so that's why I have some duplicate code to set the C++ args.  gen_doc target compiles the ngfx backend code as well when ngfx is enabled (disabled by default)

ylaala-gpsw commented on libnodegl/meson.build:
```diff
@@ -38,6 +38,11 @@ host_system = host_machine.system()
 cpu_family = host_machine.cpu_family()
 cc = meson.get_compiler('c')
 
+if cc.get_id() == 'msvc'
```
same remark about std

jmoguillansky-gpsw replied:
same comment as before: I'm using some C++20 features in ngfx, but mostly C++17

ylaala-gpsw commented on libnodegl/meson.build:
```diff
@@ -705,9 +761,17 @@ install_data(files('android/java/OnFrameAvailableListener.java'), install_dir: d
+# add_global_arguments doesn't seem to work here
+  gen_specs_cpp_args = []
+  if cc.get_id() == 'msvc'
+    gen_specs_cpp_args += ['/std:c++20']
```
same remark about std

jmoguillansky-gpsw replied:
same comment as before, add_global_arguments didn't work, so this is basically a workaround

ylaala-gpsw commented on libnodegl/backends/ngfx/util_ngfx.cpp:
```diff
@@ -0,0 +1,222 @@
+}
+
+ngfx::RenderPass* get_compat_render_pass(ngfx::GraphicsContext* ctx,
+        const struct rendertarget_desc& desc) {
```
Is this the way other class are formatted?

jmoguillansky-gpsw replied:
updated: applied clang-format to all the code under backend/ngfx.  Unfortunately the coding style in backends/ngfx doesn't conform perfectly with the rest of the code in node.gl, but this coding style isn't well defined.  We can't manually update all the code under backends/ngfx because this would have too much impact on the schedule.  What we can do is we can use a tool (I'm currently using clang-format) to automatically update the coding style but we need to agree to the coding style and define it in a clang-format file.  This will greatly speed up efficiency and accelerate the schedule.

ylaala-gpsw commented on libnodegl/backends/ngfx/util_ngfx.cpp:
```diff
@@ -0,0 +1,222 @@
+/*
```
We should probably try to stay closer of existing file naming of nodegl:
* https://github.com/gopro/gopro-lib-node.gl/blob/master/libnodegl/backends/vk/vkutils.c

jmoguillansky-gpsw replied:
If Matthieu requests this change I will rename.  
But this seems to be consistent, for example:
command_vk.c

ylaala-gpsw commented on libnodegl/backends/ngfx/texture_ngfx.h:
```diff
@@ -0,0 +1,46 @@
+struct texture *ngli_texture_ngfx_create(struct gpu_ctx *gpu_ctx);
+
+int ngli_texture_ngfx_init(struct texture *s,
+                      const struct texture_params *params);
```
format?

jmoguillansky-gpsw replied:
updated: align arguments after open bracket

ylaala-gpsw commented on libnodegl/backends/ngfx/util_ngfx.cpp:
```diff
@@ -0,0 +1,222 @@
+using namespace ngfx;
+using namespace std;
+
+FilterMode to_ngfx_filter_mode(int filter) {
```
ident

jmoguillansky-gpsw replied:
updated.  newline before opening bracket

ylaala-gpsw commented on libnodegl/backends/ngfx/texture_ngfx.cpp:
```diff
@@ -0,0 +1,97 @@
+    delete s_priv->v;
+    ngli_freep(sp);
+}
+
```
extra line?

jmoguillansky-gpsw replied:
they also have trailing newlines at end of file (see texture_vk.c).

ylaala-gpsw commented on libnodegl/backends/ngfx/texture_ngfx.cpp:
```diff
@@ -0,0 +1,97 @@
+    return 0;
+}
+
+void ngli_texture_ngfx_freep(struct texture **sp) {
```
ident

jmoguillansky-gpsw replied:
fixed

ylaala-gpsw commented on libnodegl/backends/ngfx/rendertarget_ngfx.cpp:
```diff
@@ -0,0 +1,94 @@
+        return NULL;
+    s->parent.gpu_ctx = gpu_ctx;
+    return (struct rendertarget *)s;
+}
```
extra line + indent

jmoguillansky-gpsw replied:
indent fixed.  extra line: can you clarify?  seems the same as existing files (see texture_vk.c as example)

ylaala-gpsw commented on libnodegl/backends/ngfx/hwmap_videotoolbox_darwin_ngfx.mm:
```diff
@@ -0,0 +1,175 @@
+        ngfx::MTLTexture *texture_ngfx = (ngfx::MTLTexture *)plane_ngfx->v;
+        texture_ngfx->v = mtl_texture;
+
+        //TODO: call CFRelease(texture_ref) when texture is rendered
```
Is this a leak we need to fix?

ylaala-gpsw commented on libnodegl/backends/ngfx/gpu_ctx_ngfx.cpp:
```diff
@@ -0,0 +1,646 @@
+        default_rt_desc.depth_stencil.format = to_ngli_format(graphics_context->depthStencilFormat);
+    }
+
+    // TODO: query programmatically
```
ngfx v2, we need to keep track of it @wmacdonald-gpsw

jmoguillansky-gpsw replied:
some backends like metal don't provide an API to query these values, they're defined in the spec.  I chose values that seem reasonable based on the specification

ylaala-gpsw reviewed (commented):
@jmoguillansky-gpsw great work, mostly some syntax / format cleaning. I do agree we should split the last commit. Multitple commits including 1 to introduce ngfx files and 1 to provide ngfx build support

jmoguillansky-gpsw commented on tests/shape.py:
```diff
@@ -355,42 +355,53 @@ def cropboard(cfg, dim_clr=3, dim_cut=9):
     render = ngl.Render(geometry, program, nb_instances=2)
-    render.update_instance_attributes(matrix=matrices)
+    render.update_instance_attributes(
+        matrix_0=matrices[0], matrix_1=matrices[1], matrix_2=matrices[2], matrix_3=matrices[3]
```
If you look at VkVertexInputAttributeDescription: https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VkVertexInputAttributeDescription.html

there's no format for 4x4 matrix, they emulate support for mat4 attribute using four vec4 attributes, each one with R32G32B32A32_SFLOAT format.  
You can confirm this if you run the vulkan backend or capture with renderdoc.

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -206,6 +236,38 @@ def real_decorator(block_func):
     return real_decorator
 
 
+def _get_cmake_setup_options(
```
this one is a generic function for doing cmake setup, it's not specific to shaderc

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -242,16 +304,83 @@ def _renderdoc_install(cfg):
-        renderdoc_dir = cfg.externals[_RENDERDOC_ID]
-        nodegl_opts += [f"-Drenderdoc_dir={renderdoc_dir}"]
+        nodegl_debug_opts += [f"-Ddebug_opts={debug_opts}"]
+        if "gpu_capture" in cfg.args.debug_opts:
```
this change is probably not strictly necessary.  I just split up the variable into two variables so that the code was easier to manage, but I can reuse the same variable name to minimize the changes

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -242,16 +304,83 @@ def _renderdoc_install(cfg):
-        nodegl_opts += [f"-Drenderdoc_dir={renderdoc_dir}"]
+        nodegl_debug_opts += [f"-Ddebug_opts={debug_opts}"]
+        if "gpu_capture" in cfg.args.debug_opts:
+            if _SYSTEM in ("Linux", "Windows"):
```
renderdoc is only supported on windows / linux

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -273,13 +402,15 @@ def _nodegl_setup(cfg):
     if extra_include_dirs:
         opts = ",".join(extra_include_dirs)
-        nodegl_opts += [f"-Dextra_include_dirs={opts}"]
+        nodegl_setup_opts += [f"-Dextra_include_dirs={opts}"]
```
I can remove this renaming to minimize the changes

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -462,37 +594,63 @@ def _get_make_vars(cfg):
         cfg.pkg_config_path,
         "--buildtype",
-        "debugoptimized" if debug else "release",
+        buildtype,
```
I want that change because debug lets me set breakpoints....
debugoptimized does auto inlining and loop unrolling, etc and doesn't support debugging / setting breakpoints very well.  If they reject this change I can revert it and just modify it locally

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -462,37 +594,63 @@ def _get_make_vars(cfg):
         # Workaround Debian/Ubuntu bug; see https://github.com/mesonbuild/meson/issues/5925
         meson_setup += ["--libdir=lib"]
 
+    if _SYSTEM == "Windows":
```
this is an msvc-specific option.  
I explain the reason in the comments...
some third party libs don't support "multithreaded debug DLL" (they're available in release, not debug mode), and msvc doesn't support mixing different runtime library configs.

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -579,7 +737,25 @@ def __init__(self, args):
+        if args.enable_ngfx_backend:
+            self.ngfx_graphics_backend = os.getenv(
+                "NGFX_GRAPHICS_BACKEND",
+                "NGFX_GRAPHICS_BACKEND_" + {"Windows": "DIRECT3D12", "Linux": "VULKAN", "Darwin": "METAL"}[_SYSTEM],
```
this is an environment variable used by ngfx, I use "DIRECT3D12" in the ngfx CMakelists file.
if we want to change to D3D12 we need to update ngfx as well....

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -579,7 +737,25 @@ def __init__(self, args):
+
+        self.cmake_generator = os.getenv(
+            "CMAKE_GENERATOR",
+            {"Windows": "Visual Studio 16 2019", "Linux": "Ninja", "Darwin": "Xcode", "MinGW": "Ninja"}[_SYSTEM],
```
exactly

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -579,7 +737,25 @@ def __init__(self, args):
+
         if _SYSTEM == "Windows":
+            # Set Windows SDK Version
+            self.cmake_system_version = os.getenv("CMAKE_SYSTEM_VERSION", "10.0.22000.0")
```
can override via env variable

jmoguillansky-gpsw commented on libnodegl/math_utils.h:
```diff
@@ -28,6 +28,10 @@
 #define M_PI 3.14159265358979323846
 #endif
 
+#if defined _WIN32 && !defined(restrict)
```
updated.  Added support for mac os too (darwin)

jmoguillansky-gpsw commented on libnodegl/meson.build:
```diff
@@ -243,10 +243,13 @@ hosts_cfg = {
   dependency('libsxplayer', version: '>= 9.12.0'),
   dependency('threads'),
 ]
+if host_system == 'linux'
```
yeah, I tested on Mac OS (Darwin), and these libraries aren't used

jmoguillansky-gpsw commented on libnodegl/meson.build:
```diff
@@ -243,10 +243,13 @@ hosts_cfg = {
 
 lib_deps = [
-  cc.find_library('m', required: false),
+
```
updated

jmoguillansky-gpsw commented:
> @jmoguillansky-gpsw great work, mostly some syntax / format cleaning. I do agree we should split the last commit. Multitple commits including 1 to introduce ngfx files and 1 to provide ngfx build support

Thanks @ylaala-gpsw , updated.  Split into multiple commits

jmoguillansky-gpsw commented on tests/shape.py:
```diff
@@ -355,42 +355,53 @@ def cropboard(cfg, dim_clr=3, dim_cut=9):
+    matrices = convert_mat4_attr(
+        [
+            # fmt: off
+        1.0, 0.0, 0.0, 0.0,
```
updated

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -134,22 +147,34 @@ def _download_extract(dep_item):
     name, dep = dep_item
 
-    version = dep["version"]
+    version = dep["version"] if "version" in dep else ""
```
updated

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -134,22 +147,34 @@ def _download_extract(dep_item):
+    version = dep["version"] if "version" in dep else ""
     url = dep["url"].replace("@VERSION@", version)
-    chksum = dep["sha256"]
+    chksum = dep.get("sha256")
```
updated

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -134,22 +147,34 @@ def _download_extract(dep_item):
     url = dep["url"].replace("@VERSION@", version)
-    chksum = dep["sha256"]
+    chksum = dep.get("sha256")
+    is_git_repo = url.endswith(".git")
```
updated

jmoguillansky-gpsw commented on libnodegl/math_utils.h:
```diff
@@ -28,8 +28,12 @@
 #endif
 
-#if defined _WIN32 && !defined(restrict)
+#if !defined(restrict)
```
TODO: move to separate file

jmoguillansky-gpsw commented on configure.py:
```diff
@@ -273,13 +402,15 @@ def _nodegl_setup(cfg):
 
-    return ["$(MESON_SETUP) -Drpath=true " + _cmd_join(*nodegl_opts, "libnodegl", op.join("builddir", "libnodegl"))]
+    return [
+        "$(MESON_SETUP) -Drpath=true " + _cmd_join(*nodegl_setup_opts, "libnodegl", op.join("builddir", "libnodegl"))
```
updated.  renamed builddir though to env variable
